scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchGenerator.java

```diff
@@ -0,0 +1,536 @@
+
+    private final T resource;
+    private final T original;
+
+    private final Schema schema;
+
+    private final SchemaRegistry schemaRegistry;
+
+    private final Map<Schema.Attribute, Integer> addRemoveOffsetMap = new HashMap<>();
+
+    private UpdateRequest(T original, T resource, SchemaRegistry schemaRegistry) {
+      this.schemaRegistry = schemaRegistry;
+      this.original = original;
+      this.resource = resource;
+      this.schema = schemaRegistry.getSchema(original.getBaseUrn());
+    }
+
+    public List<PatchOperation> getPatchOperations() {
+      try {
+          return createPatchOperations();
+        } catch (IllegalArgumentException | IllegalAccessException | JsonProcessingException e) {
+          throw new IllegalStateException("Error creating the patch list", e);
+        }
+    }
+
+    private void sortMultiValuedCollections(Object obj1, Object obj2, AttributeContainer ac) throws IllegalArgumentException {
+      for (Schema.Attribute attribute : ac.getAttributes()) {
+        Schema.AttributeAccessor accessor = attribute.getAccessor();
+        if (attribute.isMultiValued()) {
+          @SuppressWarnings("unchecked")
+          List<Object> collection1 = obj1 != null ? (List<Object>) accessor.get(obj1) : null;
+          @SuppressWarnings("unchecked")
+          List<Object> collection2 = obj2 != null ? (List<Object>) accessor.get(obj2) : null;
+
+          Set<Object> priorities = findCommonElements(collection1, collection2);
+          PrioritySortingComparator prioritySortingComparator = new PrioritySortingComparator(priorities);
+          if (collection1 != null) {
+            Collections.sort(collection1, prioritySortingComparator);
+          }
+
+          if (collection2 != null) {
+            Collections.sort(collection2, prioritySortingComparator);
+          }
+        } else if (attribute.getType() == Schema.Attribute.Type.COMPLEX) {
+          Object nextObj1 = obj1 != null ? accessor.get(obj1) : null;
+          Object nextObj2 = obj2 != null ? accessor.get(obj2) : null;
+          sortMultiValuedCollections(nextObj1, nextObj2, attribute);
+        }
+      }
+    }
+
+    private Set<Object> findCommonElements(List<Object> list1, List<Object> list2) {
+      if (list1 == null || list2 == null) {
+        return Collections.emptySet();
+      }
+
+      Set<Object> set1 = new HashSet<>(list1);
+      Set<Object> set2 = new HashSet<>(list2);
+
+      set1 = set1.stream().map(PrioritySortingComparator::getComparableValue).collect(Collectors.toSet());
+      set2 = set2.stream().map(PrioritySortingComparator::getComparableValue).collect(Collectors.toSet());
+
+      set1.retainAll(set2);
+      return set1;
+    }
+
+    /**
+     * There is a know issue with the diffing tool that the tool will attempt to move empty arrays. By
+     * nulling out the empty arrays during comparison, this will prevent that error from occurring. Because
+     * deleting requires the parent node
```

Comment:
```suggestion
     * There is a known issue with the diffing tool that the tool will attempt to move empty arrays. By
     * nulling out the empty arrays during comparison, this will prevent that error from occurring because
     * deleting requires the parent node.
```